raiden-ts/src/helpers.ts

```diff
@@ -341,3 +344,51 @@ export const isValidUrl = (url: string): boolean => {
       : /^(?:(http|https):\/\/)?([^\s\/$.?#&"']+\.)*[^\s\/$?#&"']+(?:(\d+))*$/;
   return regex.test(url);
 };
+
+/**
+ * Construct entire ContractsInfo using UserDeposit contract address as entrypoint
+ *
+ * @param provider - Ethers provider to use to fetch contracts data
+ * @param userDeposit - UserDeposit contract address as entrypoint
+ * @returns contracts info, with blockNumber as block of first registered tokenNetwork
+ */
+export async function fetchContractsInfo(
+  provider: JsonRpcProvider,
+  userDeposit: Address,
+): Promise<ContractsInfo> {
+  const userDepositContract = UserDepositFactory.connect(userDeposit, provider);
+
+  const oneToN = (await userDepositContract.one_to_n_address()) as Address;
+  const monitoringService = (await userDepositContract.msc_address()) as Address;
+  const monitoringServiceContract = MonitoringServiceFactory.connect(monitoringService, provider);
+
+  const tokenNetworkRegistry = (await monitoringServiceContract.token_network_registry()) as Address;
+  const tokenNetworkRegistryContract = TokenNetworkRegistryFactory.connect(
+    tokenNetworkRegistry,
+    provider,
+  );
+
+  const secretRegistry = (await tokenNetworkRegistryContract.secret_registry_address()) as Address;
+  const serviceRegistry = (await monitoringServiceContract.service_registry()) as Address;
+
+  const logs = await provider.getLogs({
+    ...tokenNetworkRegistryContract.filters.TokenNetworkCreated(null, null),
```

Comment:
Why does actually the token network contracts define the first block? Couldn't the user deposit contract be older and still includes interesting information to fetch even before the first network got deployed? :thinking: 
Sorry if this is a stupid question. But I know most-likely nothing about the contract architecture so far. :see_no_evil:

Reply:
Because this contract is the most important one, and also one of the first. The `UserDeposit` and other services contracts are deployed only later in time (even because they must contain an address pointer to the TokenNetworkRegistry, therefore need to be known at instantiation time). The first deployed contract is actually the `SecretRegistry` (since `TokenNetworkRegistry` contains a pointer to it) and then the `TokenNetworkRegistry`, but only the later is guaranteed to have an event emitted early (the first registered `TokenNetwork`, through the `TokenNetworkCreated` event), and therefore it's the event we must scan for.